flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistry.java

```diff
@@ -66,10 +67,25 @@ StreamStateHandle registerReference(
     /**
      * Register given shared states in the registry.
      *
+     * <p>NOTE: For state from checkpoints from other jobs or runs (i.e. after recovery), please use
+     * {@link #registerAllAfterRestored(CompletedCheckpoint, RestoreMode)}
+     *
      * @param stateHandles The shared states to register.
      * @param checkpointID which uses the states.
      */
     void registerAll(Iterable<? extends CompositeStateHandle> stateHandles, long checkpointID);
 
+    /**
+     * Set the lowest checkpoint ID below which no state is discarded, inclusive.
+     *
+     * <p>After recovery from an incremental checkpoint, its state should NOT be discarded, even if
+     * {@link #unregisterUnusedState(long) not used} anymore (unless recovering in {@link
+     * RestoreMode#CLAIM CLAIM} mode).
+     *
+     * <p>This should hold for both cases: when recovering from that initial checkpoint; and from
+     * any subsequent checkpoint derived from it.
+     */
+    void registerAllAfterRestored(CompletedCheckpoint checkpoint, RestoreMode mode);
```

Comment:
Just as what I asked in the above, can we ensure this method only be called during restore from checkpoint/savepoint?

Reply:
I don't think it's possible.
If the above call is removed (from `SharedStateRegistryFactory`),
and only the call from `CheckpointCoordinator` is left,

then in case of JM failover **before** taking any new checkpoints, the old no-claim-state will be discarded.

Reply:
What will happen if JM crashed after newer checkpoints taken? Does it mean we have to restore the newly created checkpoints and update the `highestNotClaimedCheckpointID` to the newer created checkpoint, which leads to the new created checkpoints not discarded anymore?

Reply:
Yes.

Reply:
You're proposing to allow discarding the state if the latest checkpoint path is different from that of `SavepointRestoreSettings.getRestorePath`, right?

I think it would again invalidate the initial checkpoint, because the latest checkpoint might still be using the initial shared state.

Reply:
I think current dilemma is caused by that we might not make the refactory of equivalence.

Previously, no matter whether JM crashed or just restore from checkpoint/savepoint. All checkpoints would be registered to incrase the reference counting (via `CompletedCheckpoint#registerSharedStatesAfterRestored`), and the checkpoint property of `discardOnSubume` could help whether to discard this registered checkpoint to decrease the reference on subsume. 
However, this changed after FLINK-24611. Previously, the unregister phase for SharedStateReigstry does not depend on the checkpoint id relationship, we just unregister the checkpoints independently and not consider which checkpoint id is larger. However, after FLINK-24611, this changed and I think we could persist the restored id in some place to mitigate such regression.

Reply:
> state left-over in case of JM failover is not a regression 

I think this is only ture for JM crash during on-going checkpoint discard. In other cases, this is a regression, although we had to say that the probability of JM crash is really small.

Reply:
> Previously, no matter whether JM crashed or just restore from checkpoint/savepoint. All checkpoints would be registered to incrase the reference counting (via CompletedCheckpoint#registerSharedStatesAfterRestored)

If the initial checkpoints were replaced (subsumed) by the new ones; then `discardOnSubume` property would be lost, wouldn't it?
Then the new JM would load the new checkpoints with `discardOnSubume=true` and after some time discard the initial shared state.

Reply:
I think your analysis is right before Flink-1.15, and Flink actually did not give exact defination on the restore mode before Flink-1.15.

How about persisting the restored checkpoint id in the checkpoint store and update the restored checkpoint id on each time checkpoint coordinator to call restoreSavepoint?

Reply:
In general, we should persist the restore mode and restored checkpoint id in the checkpoint store. With these information, we can make the restore and discard phase more safe.

Reply:
I agree that a restored checkpoint ID should be persisted durably somewhere. 
And it can also be subsequent checkpoints themselves. I think we need some more analysis to chose the right option and implement correctly. 

And because this is not strictly a regression, I strongly prefer to fix it separately and also not block the release.
Another aspect is that we should address this in 1.14 too (for `LEGACY` mode).

WDYT?

Reply:
I think the issue deserves its own ticket regardless of when we decide to address it, so I've created FLINK-27114 to track it.